File: beatmap/parser/cparser/omp.c
/*
 *  Copyright (©) 2015 Lucas Maugère, Thomas Mijieux
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */

#include <stdio.h>
#include <stdlib.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <unistd.h>
#include <string.h>
#include <regex.h>
#include <string.h>

#include <visibility.h>

#include "util/list/list.h"
#include "util/hashtable/hashtable.h"

#include "hs/hitsound.h"
#include "map/map.h"
#include "combocolor/combocolor.h"
#include "ho/hit_object.h"
#include "tp/timing_point.h"

#include "omp.h"
#include "re.h"

#define LINE_SIZE  8096

struct h_entry {
    char *id;
    char *value;
};

static void default_parser(char *line, struct h_entry *e)
{
    char *matches[2];
    if (re_match("^ *([^ ]*) *: *(.*) *$", line, 2, matches) == 0) {
	e->id = matches[0];
	e->value = matches[1];
    }
}

static void event_subsection_parser(char *line, struct generic_entry *ge)
{

}

static void event_parser(char *line, struct generic_entry *ge)
{
    
}

#define line_is_empty(line)						\
    (line == NULL || *line == '\0' || *line == '\r' || *line == '\n')	\

#define line_is_comment(line)					\
    (line != NULL && (*line == '/' || *(line+1) == '/'))	\


__internal
struct hash_table *omp_c_parse_osu_file(FILE *f, int version)
{
    struct hash_table *sections;
    static regex_t section_delim;
    struct list *hit_objects;
    struct list *timing_points;
    char *current_section_name = NULL;
    struct hash_table *current_section = NULL;
    
    sections = ht_create(NULL);
    hit_objects = list_new(0);
    timing_points = list_new(0);
    
    ht_add_entry(sections, "HitObjects", hit_objects);
    ht_add_entry(sections, "TimingPoints", timing_points);

    char line[LINE_SIZE];
    while (fgets(line, LINE_SIZE, f) != NULL) {
	char *match[1];
	
	if (re_match("^\\[(.*)\\]", line, 1, match) == 0) {
	    free(current_section_name);
	    current_section_name = match[0];
	    if (!ht_has_entry(sections, current_section_name)) {
		current_section = ht_create(NULL);
		ht_add_entry(sections, current_section_name, current_section);
	    } else {
		ht_get_entry(sections, current_section_name, &current_section);
	    }
	    continue;
	}

	if (!current_section_name ||
	    line_is_empty(line) ||   line_is_comment(line))
	    continue;
	
	if (!strcmp(current_section_name, "Events")) {
	    // parse subsection and event in a similar manner
	    // ev_c_parse() ...
	} else if (!strcmp(current_section_name, "HitObjects")) {
	    struct hit_object *ho = calloc(sizeof(*ho), 1);
	    ho_c_parse(line, ho, version);
	    list_add(hit_objects, ho);
	} else if (!strcmp(current_section_name, "TimingPoints")) {
	    struct timing_point *tp = malloc(sizeof(*tp));
	    tp_c_parse(line, tp);
	    list_add(timing_points, tp);
	} else {
	    struct h_entry e;
	    default_parser(line, &e);
	    ht_add_entry(current_section, e.id, e.value);
	    free(e.id);
	}
    }

    free(current_section);
    regfree(&section_delim);
    return sections;
}

File: map/py/omp.py
#!/usr/bin/env python2
# -*- encoding: utf-8 -*-

#  Copyright (©) 2015 Lucas Maugère, Thomas Mijieux
#
#  Licensed under the Apache License, Version 2.0 (the "License");
#  you may not use this file except in compliance with the License.
#  You may obtain a copy of the License at
#
#      http://www.apache.org/licenses/LICENSE-2.0
#
#  Unless required by applicable law or agreed to in writing, software
#  distributed under the License is distributed on an "AS IS" BASIS,
#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
#  See the License for the specific language governing permissions and
#  limitations under the License.

import sys, re
from omp_ho import parse_hitobject
from omp_tp import parse_timingpoint
from omp_coco import parse_colour
from omp_sb import parse_event
from omp_util import Int, Float

def parse(osufilename):
    data = {}
    f = open(osufilename);
    ver = f.readline()
    m = re.match(u"^(\u00EF\u00BB\u00BF)?osu file format v([\d]+)\r\n", ver)
    if m:
        if m.group(1):
            data['BOM'] = True
        else:
            data['BOM'] = False
        data['version'] = int(m.group(2))
    else:
        print "Cannot determine \"{0}\" format!".format(osufilename)
        return

    for line in f:
        line = line.rstrip()
        a = re.match(r"^\[(.*)\]$", line)
        if a:
            current_section = a.group(1)
            if current_section in ["TimingPoints", "HitObjects",
                                   "Events", "Colours"]: 
                data[current_section] = [] # need list
            else: 
                data[current_section] = {} # need dict
            continue
        
        if line == "" or (line[0] =='/' and line[1] == '/'):
            continue # ignore comments or empty lines
        
        try:
            current_section
        except:
            current_section = None

        if current_section is not None:
            
            if current_section == "HitObjects":
                data[current_section].append(parse_hitobject(line))
                
            elif current_section == "TimingPoints":
                data[current_section].append(parse_timingpoint(line))
                
            elif current_section == "Events":
                ev = parse_event(line)
                if ev not in [None, {}]:
                    # this check is necesseary because the function build the
                    # object over calls (a single sprite can be splitted on
                    # multiple lines)
                    data[current_section].append(ev)
                    
            elif current_section == "Colours":
                data[current_section].append(parse_colour(line))

            else:
                m = re.match(r"^ *([^ ]*) *: *(.*) *$", line)
                if m:
                    keyword, value = m.group(1), m.group(2)
                    if keyword in  ["Bookmarks", "Tags"]:
                        value = map(Int, value.split(","))
                    value = Float(value)
                    data[current_section][keyword] = value

                    
    return data


if __name__ == "__main__":
    if len(sys.argv) != 2:
        print "error"
        sys.exit(1)

    data = parse(sys.argv[1])
    import pprint
    pp = pprint.PrettyPrinter(indent=0)
    pp.pprint(data)


